### PR TITLE
KAFKA-7158: Add unit test for window store range queries

### PR DESCRIPTION
**Pull request by guozhangwang**
While debugging the reported issue, I found that our current unit test lacks coverage to actually expose the underlying root cause.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java]: nit: rename method -> `shouldNotReturnDuplicatesInRanges` ?

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java]: nit: indention should be 4 spaces only?

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java]: nit: insert empty line

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java]: nit: do we need this comments? (similar below)

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java]: We should initialize `TopologyTestDriver` with a fixed mock-timestamp and use this below instead of calling `System. currentTimeMillis()`

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java]: Is it intended that all records fall into a single window / segment? What was the root cause of the bug and how does this data/timestamp pattern cover it? (cannot remember)

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java]: ack

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java]: ack

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java]: ack

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java]: ack

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java]: It is actually not the root cause: as long as we have records with somewhat different timestamps, they may trigger the issue in caching layer for range queries.

**Review by guozhangwang — COMMENTED** [review]
Addressed comments.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java]: Thanks for clarification. So it does not really matter what data we use in the test. Would there be a minimal setup that exposes the issue reliably? Just wonder if the test could be "more clear" from the code. If not, we can leave as is.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java]: Do you want to pass in `autoAdvanceMs` for record creation? Default is zero.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java]: So you want to advance `recordFactory` time too? (similar below)

**Review by mjsax — COMMENTED** [review]
Thanks for the update. Is it intended that all generated records have the same timestamp now? See inline comment.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java]: I've explicitly called `recordFactory.advanceTimeMs(10 * 1000L);` after `driver.advanceWallClockTime(10 * 1000L);`.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java]: I explicitly call `recordFactory.advanceTimeMs` now.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java]: nit: Since each record is created with `randomUUID` for the keys and values,  maybe create a method accepting an `int` for the number of desired `driver.pipeInput` calls and drive the input inside a loop?

**Review by bbejeck — APPROVED** [review]
Thanks, @guozhangwang.  I have just one minor comment otherwise LGTM.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java]: Ack.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java]: We could use nested for-loops.

**Review by mjsax — APPROVED** [review]
LGTM.

One more minor suggestions -- feel free to ignore. And feel free to merge.

**Review by bbejeck — APPROVED** [review]
LGTM

**Comment by guozhangwang** [comment]
Cherry-picked to 2.0 after merging https://github.com/apache/kafka/pull/5474.

**Pull request merged**
